tree/ntuple/v7/test/SimpleCollectionProxy.hxx

```diff
@@ -0,0 +1,89 @@
+   {
+   }
+   SimpleCollectionProxy(const SimpleCollectionProxy &) : SimpleCollectionProxy() {}
+
+   TVirtualCollectionProxy *Generate() const override { return new SimpleCollectionProxy<CollectionT>(*this); }
+   Int_t GetCollectionType() const override { return ROOT::kSTLvector; }
+   ULong_t GetIncrement() const override { return sizeof(typename CollectionT::ValueType); }
+   UInt_t Sizeof() const override { return sizeof(CollectionT); }
+   Bool_t HasPointers() const override { return kFALSE; }
+
+   TClass *GetValueClass() const override
+   {
+      if constexpr (std::is_fundamental<typename CollectionT::ValueType>::value)
+         return nullptr;
+      return TClass::GetClass(ROOT::Internal::GetDemangledTypeName(typeid(typename CollectionT::ValueType)).c_str());
+   }
+   EDataType GetType() const override
+   {
+      if constexpr (std::is_same<typename CollectionT::ValueType, char>::value)
+         return EDataType::kChar_t;
+      ;
+      if constexpr (std::is_same<typename CollectionT::ValueType, float>::value)
+         return EDataType::kFloat_t;
+      ;
+      return EDataType::kOther_t;
+   }
+
+   void PushProxy(void *objectstart) override { fObject = static_cast<CollectionT *>(objectstart); }
+   void PopProxy() override { fObject = nullptr; }
+
+   void *At(UInt_t idx) override { return &fObject->v[idx]; }
+   void Clear(const char * /*opt*/ = "") override { fObject->v.clear(); }
+   UInt_t Size() const override { return fObject->v.size(); }
+   void *Allocate(UInt_t /*n*/, Bool_t /*forceDelete*/) override { return nullptr; }
+   void Commit(void *) override {}
+   void Insert(const void *data, void *container, size_t size) override
+   {
+      auto p = static_cast<const typename CollectionT::ValueType *>(data);
+      for (size_t i = 0; i < size; ++i) {
+         static_cast<CollectionT *>(container)->v.push_back(p[i]);
+      }
+   }
+
+   TStreamerInfoActions::TActionSequence *
+   GetConversionReadMemberWiseActions(TClass * /*oldClass*/, Int_t /*version*/) override
+   {
+      return nullptr;
+   }
+   TStreamerInfoActions::TActionSequence *GetReadMemberWiseActions(Int_t /*version*/) override { return nullptr; }
+   TStreamerInfoActions::TActionSequence *GetWriteMemberWiseActions() override { return nullptr; }
+
+   CreateIterators_t GetFunctionCreateIterators(Bool_t /*read*/ = kTRUE) override { return nullptr; }
+   CopyIterator_t GetFunctionCopyIterator(Bool_t /*read*/ = kTRUE) override { return nullptr; }
+   Next_t GetFunctionNext(Bool_t /*read*/ = kTRUE) override { return nullptr; }
+   DeleteIterator_t GetFunctionDeleteIterator(Bool_t /*read*/ = kTRUE) override { return nullptr; }
+   DeleteTwoIterators_t GetFunctionDeleteTwoIterators(Bool_t /*read*/ = kTRUE) override { return nullptr; }
+};
+} // namespace
+
+namespace ROOT::Experimental {
+template <>
+struct IsCollectionProxy<StructUsingCollectionProxy<char>> : std::true_type {
```

Comment:
Hm, can you expand on why this is needed? If I understand the comment in RField.hxx correctly, that means that users have to mark all their collection proxy classes in the same way. Hm, I wonder how much of a limitation that is an if it can be avoided.

Reply:
Yes, the point here is that we can only tell at run time if a class has an associated collection proxy.

For compile time, in the first iteration of the PR we had an extra template argument that acted as a "tag" to differentiate the RField specialization for classes with an associated collection proxy (inherits `RCollectionClassField`) from the RField primary template definition (`RClassField`-derived), as in:
```c++
auto field = std::make_unique<RField<MyClass>>("klass");
// vs
auto otherField = std::make_unique<RField<MyClass, ROOT::Experimental::TagIsCollectionProxy>>("klass"); // Using collection proxy
```

That is convenient only for non-nested types, i.e. it doesn't work with, e.g. `RField<std::vector<MyClass>, ROOT::Experimental::TagIsCollectionProxy>`, as the tag is not forwarded to the instantiation of the inner RField (that for the value type of the vector).  I considered the following two possible solutions:
- A wrapper type (much like https://github.com/root-project/root/blob/master/tree/ntuple/v7/inc/ROOT/RNTupleUtil.hxx#L49), that helps to differentiate both cases.  There we would have:
```c++
auto field = std::make_unique<RField<RProxiedCollection<MyClass>>>("klass"); // Using collection proxy
```
- A helper `IsCollectionProxy<T>` type, that can be used in a similar way to those in the `<type_traits>` header.  I found this more convenient and is the implemented thing in this PR.  Here, classes can be marked as a collection proxy with either of the following two forms (whichever is more convenient for the user):
```c++
template <>
struct IsCollectionProxy<MyClass> : std::true_type {};
```
or by adding a member type to the class as follows:
```c++
class MyClass {
public:
   using IsCollectionProxy = std::true_type;
};

Of course, there is another possible solution which is to have a single `RClassField` that implements both the regular-class and the collection-proxy behaviors, and always chooses appropriately at run time.  I found that less clean and probably has more overhead, as most probably it involves an additional branch + call in each of the member functions.

Reply:
Many thanks for the explanation!  I think I got it.  Could you add the text from this comment as a code comment?  I think that would very much help the next developer.

One more question: If a user forgets to set the `IsCollectionProxy` trait, what would happen if they try to create a field of such a type with an attached collection proxy? I guess RNTuple would go with `RClassField` and then throw an exception because it unexpectedly found a collection proxy attached to the class, right? I think that would work well, overall.

Reply:
Done!

> If a user forgets to set the `IsCollectionProxy` trait, what would happen if they try to create a field of such a type with an attached collection proxy? I guess RNTuple would go with `RClassField` and then throw an exception because it unexpectedly found a collection proxy attached to the class, right? I think that would work well, overall.

Exactly!

> For RFieldBase::Create("type as a string"), we don't need the type trait as the distinction regular class / collection proxy class is fully performed at runtime, right?

That's absolutely right.